velox/type/Filter.cpp

```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+      NegatedBigintValuesUsingBitmask::create);
+  registry.Register("FloatingPointRange", AbstractRange::create);
+  registry.Register("BytesRange", BytesRange::create);
+  registry.Register("NegatedBytesRange", NegatedBytesRange::create);
+  registry.Register("BytesValues", BytesValues::create);
+  registry.Register("BigintMultiRange", BigintMultiRange::create);
+  registry.Register("NegatedBytesValues", NegatedBytesValues::create);
+  registry.Register("MultiRange", MultiRange::create);
+}
+
+folly::dynamic Filter::serializeBase(std::string_view name) const {
+  folly::dynamic obj = folly::dynamic::object;
+  obj["name"] = name;
+  obj["nullAllowed"] = nullAllowed_;
+  obj["deterministic"] = deterministic_;
+  obj["kind"] = filterKinds(kind_);
+  return obj;
+}
+
+folly::dynamic AlwaysFalse::serialize() const {
+  return Filter::serializeBase("AlwaysFalse");
+}
+
+FilterPtr AlwaysFalse::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<AlwaysFalse>();
+}
+
+folly::dynamic AlwaysTrue::serialize() const {
+  return Filter::serializeBase("AlwaysTrue");
+}
+
+FilterPtr AlwaysTrue::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<AlwaysTrue>();
+}
+
+folly::dynamic IsNull::serialize() const {
+  return Filter::serializeBase("IsNull");
+}
+
+FilterPtr IsNull::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<IsNull>();
+}
+
+folly::dynamic IsNotNull::serialize() const {
+  return Filter::serializeBase("IsNotNull");
+}
+
+FilterPtr IsNotNull::create(const folly::dynamic& /*obj*/) {
+  return std::make_unique<IsNotNull>();
+}
+
+folly::dynamic BoolValue::serialize() const {
+  auto obj = Filter::serializeBase("BoolValue");
+  obj["value"] = value_;
+  return obj;
+}
+
+FilterPtr BoolValue::create(const folly::dynamic& obj) {
+  auto value = obj["value"].asBool();
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  return std::make_unique<BoolValue>(value, nullAllowed);
+}
+
+bool BoolValue::testingEquals(const Filter& other) const {
+  auto otherBoolValue = dynamic_cast<const BoolValue*>(&other);
+  return otherBoolValue != nullptr && Filter::testingEquals(other) &&
+      (value_ == otherBoolValue->value_);
+}
+
+folly::dynamic BigintRange::serialize() const {
+  auto obj = Filter::serializeBase("BigintRange");
+  obj["lower"] = lower_;
+  obj["upper"] = upper_;
+  return obj;
+}
+
+FilterPtr BigintRange::create(const folly::dynamic& obj) {
+  auto lower = obj["lower"].asInt();
+  auto upper = obj["upper"].asInt();
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  return std::make_unique<BigintRange>(lower, upper, nullAllowed);
+}
+
+bool BigintRange::testingEquals(const Filter& other) const {
+  auto otherBigintRange = dynamic_cast<const BigintRange*>(&other);
+  return otherBigintRange != nullptr && Filter::testingEquals(other) &&
+      (lower_ == otherBigintRange->lower_) &&
+      (upper_ == otherBigintRange->upper_) &&
+      (isSingleValue_ == otherBigintRange->isSingleValue_);
```

Comment:
Is this check necessary? Assuming isSingleValue_ is derived from lower and upper_, this check will always pass, no?